### PR TITLE
executor: Fix a bug in 'INSERT ... ON DUPLICATE KEY UPDATE'

### PR DESCRIPTION
**Pull request by bb7133**
### What problem does this PR solve?
Fix a bug in 'INSERT ... ON DUPLICATE KEY UPDATE' when unique key are updated, for example:

```
mysql> select * from t1;
+------+------+------+
| id   | a    | b    |
+------+------+------+
|    1 |    1 |    1 |
|    2 |    1 |    2 |
|    3 |    3 |    1 |
+------+------+------+
3 rows in set (0.00 sec)

mysql> create table t2(a int key, b int, unique(b));
Query OK, 0 rows affected (0.08 sec)

mysql> insert into t2 select a, b from t1 order by id on duplicate key update a=t1.a, b=t1.b;
Query OK, 6 rows affected (0.01 sec)

mysql> select * from t2;
+---+------+
| a | b    |
+---+------+
| 3 |    1 |
+---+------+
1 row in set (0.00 sec)
``` 
, which is incorrect. We should expect the following result instead:
```
mysql> select * from t2;
+---+------+
| a | b    |
+---+------+
| 1 |    2 |
| 3 |    1 |
+---+------+
2 rows in set (0.00 sec)
```

After this PR, it will be fixed.

### What is changed and how it works?
In `InsertExec`'s `updateDupKeyValues` function, `dupKVs`(a `map` that holds existing primary keys and unique keys for duplication detection) is updated by removing keys from the _inserting row_ first, then adding keys from the _updated row_. But we should remove keys from the _old row_ instead, because _old row_ is the one that to be replaced.

### Check List

Tests
 - Integration test

Code changes
 - Has exported function/method change

Side effects
 - No side effect

Related changes
 - Nothing else

### CODE REVIEW

**Comment by bb7133** [comment]
@jackysp @zz-jason PTAL, thanks!

**Comment by zz-jason** [comment]
@XuHuaiyu @winoros PTAL

**Comment by zz-jason** [comment]
LGTM

**Comment by shenli** [comment]
@jackysp PTAL

File: executor/insert.go [hunk 1]
```diff
@@ -215,15 +215,22 @@ func (e *InsertExec) doDupRowUpdate(handle int64, oldRow []types.Datum, newRow [
...
+func (e *InsertExec) updateDupKeyValues(oldHandle int64, newHandle int64,
+	handleChanged bool, oldRow []types.Datum, updatedRow []types.Datum) error {
 	// There is only one row per update.
 	fillBackKeysInRows, err := e.getKeysNeedCheck(e.ctx, e.Table, [][]types.Datum{updatedRow})
 	if err != nil {
 		return errors.Trace(err)
 	}
 	// Delete old keys and fill back new key-values of the updated row.
-	e.deleteDupKeys(row)
+	cleanupRows, err := e.getKeysNeedCheck(e.ctx, e.Table, [][]types.Datum{oldRow})
```
Comment (jackysp) [inline executor/insert.go]: We could use a function to handle this part and https://github.com/pingcap/tidb/blob/9070bb6b86b27e59f35ebc2f984e369d3c5a115d/executor/replace.go#L73
together.

Comment (bb7133) [reply executor/insert.go]: Thanks for the review @jackysp, I plan to define a function for this and put it in the base struct `InsertValues`, what do u think?

Comment (jackysp) [reply executor/insert.go]: I think to put it into `batchChecker` is better. Thanks for your contribution, @bb7133 !

**Comment by bb7133** [comment]
Comments addressed, PTAL @jackysp , thanks!

**Comment by bb7133** [comment]
Thank you @jackysp

**Pull request merged**